File: backend/services/world_reset_service.py
"""
World reset service for restoring worlds to initial state.

This module handles saving and loading the initial state snapshot
used for the "reset world" functionality.
"""

import json
import logging
from datetime import datetime
from pathlib import Path
from typing import Any, Optional

from services.world_service import WorldService

logger = logging.getLogger("WorldResetService")

# Filename for initial state snapshot
INITIAL_STATE_FILE = "_initial.json"


class WorldResetService:
    """Handles world reset operations."""

    @staticmethod
    def get_initial_state_path(world_name: str) -> Path:
        """Get the path to a world's initial state file."""
        return WorldService.get_world_path(world_name) / INITIAL_STATE_FILE

    @staticmethod
    def load_initial_state(world_name: str) -> Optional[dict[str, Any]]:
        """
        Load initial state from _initial.json.

        Args:
            world_name: Name of the world

        Returns:
            Initial state dict or None if not found
        """
        initial_file = WorldResetService.get_initial_state_path(world_name)

        if not initial_file.exists():
            logger.warning(f"No {INITIAL_STATE_FILE} found for world '{world_name}'")
            return None

        try:
            with open(initial_file, "r", encoding="utf-8") as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError) as e:
            logger.error(f"Failed to load {INITIAL_STATE_FILE}: {e}")
            return None

    @staticmethod
    def save_initial_state(world_name: str, initial_state: dict[str, Any]) -> bool:
        """
        Save initial state to _initial.json.

        Args:
            world_name: Name of the world
            initial_state: State to save (starting_location, initial_stats, initial_inventory)

        Returns:
            True if successful, False otherwise
        """
        initial_file = WorldResetService.get_initial_state_path(world_name)

        try:
            with open(initial_file, "w", encoding="utf-8") as f:
                json.dump(initial_state, f, ensure_ascii=False, indent=2)
            logger.info(f"Saved initial state for world '{world_name}'")
            return True
        except IOError as e:
            logger.error(f"Failed to save {INITIAL_STATE_FILE}: {e}")
            return False

    @staticmethod
    def has_initial_state(world_name: str) -> bool:
        """
        Check if a world has saved initial state.

        Args:
            world_name: Name of the world

        Returns:
            True if _initial.json exists
        """
        return WorldResetService.get_initial_state_path(world_name).exists()

    @staticmethod
    def create_initial_state_snapshot(
        starting_location: str,
        initial_stats: dict[str, Any],
        initial_inventory: list[dict[str, Any]],
        initial_game_time: dict[str, int] | None = None,
    ) -> dict[str, Any]:
        """
        Create an initial state snapshot dict.

        Args:
            starting_location: Name of the starting location
            initial_stats: Initial stat values
            initial_inventory: Initial inventory items
            initial_game_time: Initial game time (hour, minute, day)

        Returns:
            Initial state dict ready to be saved
        """
        snapshot = {
            "starting_location": starting_location,
            "initial_stats": initial_stats,
            "initial_inventory": initial_inventory,
            "captured_at": datetime.utcnow().isoformat() + "Z",
        }
        if initial_game_time:
            snapshot["initial_game_time"] = initial_game_time
        return snapshot

File: backend/src/services/world-reset-service.ts
/**
 * The initial-state snapshot — `worlds/{name}/_initial.json`. Written once at
 * the end of onboarding, read only by the reset route.
 *
 * **A missing snapshot and an empty one are different things.** A world with no
 * file cannot be reset; a character who started with nothing has a file of
 * empty collections and resets fine — so `loadInitialState` returning `null`
 * means "absent or unreadable", never "empty".
 */

import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'

import { MtimeCache, WorldService } from './world-service'
import type { InventoryEntry } from '../domain/player-rules'
import { getLogger } from '../infrastructure/logging/logger'
import type { GameTime } from './player-service'

const logger = getLogger('WorldResetService')

const INITIAL_STATE_FILE = '_initial.json'

/** Keys are the on-disk names. `initial_game_time` is absent when unset. */
export interface InitialStateSnapshot {
  /** Location *folder* name the character starts in. */
  starting_location: string
  initial_stats: Record<string, number>
  initial_inventory: InventoryEntry[]
  captured_at: string
  initial_game_time?: GameTime
}

export interface InitialStateInput {
  startingLocation: string
  initialStats: Record<string, number>
  initialInventory: InventoryEntry[]
  initialGameTime?: GameTime | null
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

// UTC with a trailing `Z`; the last three digits are padding, never parsed.
function utcIsoNow(now: Date = new Date()): string {
  // `toISOString()` already ends in `Z`; drop it before padding.
  return `${now.toISOString().slice(0, -1)}000Z`
}

export class WorldResetService {
  private readonly worlds: WorldService

  /** `_initial.json` is read uncached; `cache` only reaches {@link WorldService}. */
  constructor(worldsDir: string, cache: MtimeCache = new MtimeCache()) {
    this.worlds = new WorldService(worldsDir, cache)
  }

  getInitialStatePath(worldName: string): string {
    return join(this.worlds.getWorldPath(worldName), INITIAL_STATE_FILE)
  }

  /** Whether the world can be reset at all. */
  hasInitialState(worldName: string): boolean {
    return existsSync(this.getInitialStatePath(worldName))
  }

  /** `null` when missing or unreadable; either way the world cannot be reset. */
  loadInitialState(worldName: string): InitialStateSnapshot | null {
    const initialFile = this.getInitialStatePath(worldName)

    if (!existsSync(initialFile)) {
      logger.warning(`No ${INITIAL_STATE_FILE} found for world '${worldName}'`)
      return null
    }

    let parsed: unknown
    try {
      parsed = JSON.parse(readFileSync(initialFile, 'utf-8'))
    } catch (error) {
      logger.error(`Failed to load ${INITIAL_STATE_FILE}: ${String(error)}`)
      return null
    }

    // Valid JSON of the wrong kind (a list, a string, `null`) is a corrupt
    // file, reported as a decode failure rather than handed to the caller.
    if (!isRecord(parsed)) {
      logger.error(`${INITIAL_STATE_FILE} for world '${worldName}' is not an object`)
      return null
    }

    return parsed as unknown as InitialStateSnapshot
  }

  /** Indented, non-ASCII unescaped, so a Korean location name stays legible. */
  saveInitialState(worldName: string, initialState: InitialStateSnapshot): boolean {
    try {
      writeFileSync(
        this.getInitialStatePath(worldName),
        JSON.stringify(initialState, null, 2),
        'utf-8',
      )
    } catch (error) {
      logger.error(`Failed to save ${INITIAL_STATE_FILE}: ${String(error)}`)
      return false
    }

    logger.info(`Saved initial state for world '${worldName}'`)
    return true
  }

  /**
   * `initial_game_time` is emitted only for a non-empty clock: absent, a reset
   * leaves the clock alone rather than winding it to midnight of day zero.
   */
  static createInitialStateSnapshot(input: InitialStateInput): InitialStateSnapshot {
    const snapshot: InitialStateSnapshot = {
      starting_location: input.startingLocation,
      initial_stats: input.initialStats,
      initial_inventory: input.initialInventory,
      captured_at: utcIsoNow(),
    }

    if (input.initialGameTime && Object.keys(input.initialGameTime).length > 0) {
      snapshot.initial_game_time = input.initialGameTime
    }

    return snapshot
  }
}
